File: mobile/composeApp/src/commonMain/kotlin/me/eyetealer/wortel/data/Dto.kt
package me.eyetealer.wortel.data

import kotlinx.serialization.Serializable
import me.eyetealer.wortel.domain.GameStatus
import me.eyetealer.wortel.domain.GuessResult

@Serializable
data class CreateGameRequest(
    val wordLength: Int = 5,
    val maxAttempts: Int = 6,
)

@Serializable
data class CreateGameResponse(
    val id: String,
    val wordLength: Int,
    val maxAttempts: Int,
)

@Serializable
data class GuessRequest(val guess: String)

@Serializable
data class RejectedGuess(val code: String, val message: String)

@Serializable
data class GameStateResponse(
    val id: String,
    val status: GameStatus,
    val attempts: List<GuessResult>,
    val remainingAttempts: Int,
    val wordLength: Int,
    val maxAttempts: Int,
    val secretWord: String? = null,
    /**
     * Set when the server accepted the request but rejected the guess
     * (unknown word, wrong length). Game state is unchanged.
     */
    val rejectedGuess: RejectedGuess? = null,
)

@Serializable
data class ApiError(val error: String, val code: String? = null)

File: supabase/functions/_shared/http/dto.ts
import { Game } from "../domain/game/Game.ts";
import { GameStatus, isTerminal } from "../domain/game/GameStatus.ts";
import { GuessResult } from "../domain/game/Guess.ts";
import { GuessRejection } from "../application/SubmitGuess.ts";

export interface CreateGameResponse {
  id: string;
  wordLength: number;
  maxAttempts: number;
}

export interface GameStateResponse {
  id: string;
  status: GameStatus;
  attempts: GuessResult[];
  remainingAttempts: number;
  wordLength: number;
  maxAttempts: number;
  secretWord?: string;
  /**
   * Present iff the last submitted guess was rejected by the dictionary
   * or for length reasons. Clients shake the UI; no game state changed.
   */
  rejectedGuess?: GuessRejection;
}

export function toCreateGameResponse(game: Game): CreateGameResponse {
  return {
    id: game.id,
    wordLength: game.wordLength,
    maxAttempts: game.maxAttempts,
  };
}

export function toGameStateResponse(
  game: Game,
  rejection?: GuessRejection,
): GameStateResponse {
  return {
    id: game.id,
    status: game.status,
    attempts: [...game.attempts],
    remainingAttempts: game.remainingAttempts,
    wordLength: game.wordLength,
    maxAttempts: game.maxAttempts,
    secretWord: isTerminal(game.status) ? game.secretWord : undefined,
    rejectedGuess: rejection,
  };
}
